File: intro/week2/vigenere.c
/*
    A program to implement vigenere's cipher.
    It take the keyword as an cmd line arg and plaintext from user
    and process it to print a scrambled version of plain text.
 */
#include <D:\CS50\cs50.h>
#include <stdio.h>
#include <ctype.h>

int main(int argc, string argv[])    //int main(int argc, string argv[])
{
    //check no of arguments = 2 or not
    if (argc != 2)
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
   // validate argument - each character should be alphabet
   int count = 0; //to count the no of letters in keyword
    for (int i = 0; argv[1][i] != '\0'; i++)
    {
        count++;
        if(!(isalpha(argv[1][i])))
        {
            printf("Usage: ./vigenere keyword\n");
            return 2;
        }
    }
    // Input plaintext
    string text = get_string("plaintext: ");
    // Cipher's Implementation --->
    for (int i = 0, j = 0; text[i] != '\0'; i++)   // iterate over every letter of plaintext, j represents                                                         keyword's letter's postion
    {
        int key;
        if(isupper(argv[1][j]))
            key = ( (int) argv[1][j] ) % ( (int) 'A' );    //find the key as a = 0,
        else if(islower(argv[1][j]))                            //b = 1, c = 2......
            key = ( (int) argv[1][j] ) % ( (int) 'a' );     //type casting from char to int
        int shift;
        if(isupper(text[i]))
        {
            shift = text[i] + key;
            while(shift > 'Z')   // wraping from above Z
                shift = ('A' - 1) + (shift % 'Z');     // implicit casting from char to int
            text[i] = shift;

                if(j < count - 1)
                    j++;
                else
                    j = 0;
        }
        else if(islower(text[i]))
        {
            shift = text[i] + key;
            while(shift > 'z')    // wraping from above z
                shift = ('a' - 1) + (shift % 'z');     // implicit casting from char to int
            text[i] = shift;

            if(j < count - 1)
                j++;
            else
                j = 0;
        }
    }
    // printing ciphertext --->
    printf("ciphertext: %s\n", text);

    return 0;
}

File: intro/week6/psets6/vigenere.py
# A program to implement vigenere's cipher.
# It take the keyword as an cmd line arg and plaintext from user
# and process it to print a scrambled version of plain text.

from sys import argv
from cs50 import get_string

def main():
    # check number of arguments
    if len(argv) != 2:
        print("Usage: python vigenere.py keyword")
        exit(1)
    
    # check whether each character is alphabet or not
    if not argv[1].isalpha():
        print("Usage: python vigenere.py keyword")
        exit(2)
    
    # input plaintext
    text = get_string("plaintext: ")

    # vigenere's cipher -->
    j = 0   # represents keywords letter
    ctext = [ch for ch in text]
    for i in range(len(text)):
        if argv[1][j].isupper():
            key = (ord(argv[1][j])) % (ord('A')) # find the key as A = a = 0... ord gives ascii value of alphabet
        elif argv[1][j].islower():
            key = (ord(argv[1][j])) % (ord('a'))
        
        if text[i].isupper():
            shift = ord(text[i]) + key
            while shift > ord('Z'): # wrapping from above 'Z'
                shift = (ord('A') - 1) + (shift % ord('Z'))
            ctext[i] = chr(shift)

            if(j < len(argv[1]) - 1):
                j += 1
            else:
                j = 0
        elif text[i].islower():
            shift = ord(text[i]) + key
            while shift > ord('z'): # wrapping from above 'Z'
                shift = (ord('a') - 1) + (shift % ord('z'))
            ctext[i] = chr(shift)

            if(j < len(argv[1]) - 1):
                j += 1
            else:
                j = 0

    # printing ciphertext
    print("ciphertext:", end=" ")
    for i in range(len(ctext)):
        print(ctext[i], end="")
    print()

if __name__ == "__main__":
    main()